File: dailyQuestion/2020-10/10-11/python/solution.py
class Solution:
    def canPartition(self, nums: [int]) -> bool:
        n = len(nums)
        if n < 2 :
            return False
        sumn = sum(nums)
        if sumn % 2 != 0 :
            return False
        target = sumn >> 1
        maxn = max(nums)
        if maxn > target :
            return False
        dp = [False] * ( target + 1)
        dp[0] = True
        for i in range(n) :
            v = nums[i]
            for j in range(target, v-1,-1) :
                dp[j] = dp[j] or dp[j-v]
        return dp[target]



if __name__ == "__main__" :
    s = Solution()
    a = s.canPartition([1,2,5])
    print(a)

File: dailyQuestion/2020-10/10-11/golang/solution.go
package main

import (
	"fmt"
)

func main() {

	result := canPartition([]int{1, 5, 11, 5})
	fmt.Println(result)
}

func canPartition(nums []int) bool {
	n := len(nums)
	if n < 2 {
		return false
	}
	sum, max := 0, 0
	for _, k := range nums {
		sum += k
		if k > max {
			max = k
		}
	}
	if sum%2 != 0 {
		return false
	}
	target := sum >> 1
	if max > target {
		return false
	}
	dp := make([]bool, target+1)
	dp[0] = true
	for i := 0; i < n; i++ {
		v := nums[i]
		for j := target; j >= v; j-- {
			dp[j] = dp[j] || dp[j-v]
		}
	}
	return dp[target]
}
